### DISCUSSION TITLE
How to globally exclude certain email addresses from the mailer?

### DISCUSSION THREAD

**Discussion by jameslkingsley** [laravel/framework#33751] [General] [answered]
How can I exclude certain email addresses from the Laravel mailer? Scenario I'm working with is when emails are marked as bounced/complaint and should be excluded in future emails. I'd like to handle this automatically where the mailer will exclude those emails for me.

Something like:
```php
Mail::setGlobalExclusions(function () {
    return arrayOfEmailAddressesToExclude();
});
```

And then if I were to use the `Mail::to()` method, it would ensure none of the excluded emails are used:
```php
public function to($users)
{
    return (new PendingMail($this))->to(
        $this->removeExclusions($users)
    );
}
```

Is this currently achievable in Laravel somehow?

**Comment by jameslkingsley** [comment]
Just had an idea that maybe I could listen for the `MailSending` event and drop the mail if it's excluded, but I'm not sure there's enough control exposed to actually do that.

**Comment by mfn** [comment]
If you control every place you send emails, you could just write a central wrapper for your `Mail::` calls which ignores certain recipients?

Sound probably too simple otherwise I guess you would have done that 🤔 😏

**Reply by jameslkingsley** [reply]
Yea I could do that I suppose, would just mean that other developers would need to be aware of said wrapper and detaches from the Laravel framework way a bit.

**Comment by jameslkingsley** [comment, marked as answer]
Think I reached an okay solution:

Registered a replacement `MailManager` in my own `MailServiceProvider`:
```php
class MailServiceProvider extends ServiceProvider
{
    public function boot()
    {
        $this->app->singleton('mail.manager', function ($app) {
            return new MailManager($app);
        });
    }
}
```

And then `MailManager` just extends the framework base, overriding the `resolve` method to use a custom `Mailer` class to adds the exclusion logic.

**Reply by jameslkingsley** [reply]
It would be better if Laravel resolved the `Mailer` class out of the container so that it could be properly overridden - at the moment it's implemented like this:
```php
protected function resolve($name)
{
    $config = $this->getConfig($name);

    if (is_null($config)) {
        throw new InvalidArgumentException("Mailer [{$name}] is not defined.");
    }

    $mailer = new Mailer(
        $name,
        $this->app['view'],
        $this->createSwiftMailer($config),
        $this->app['events']
    );

    ...
}
```

But I imagine that's largely out of scope and only applicable in rare use cases.
